### PR TITLE
:recycle: rework how `UPathIOManager` works with partitions

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

This PR is addressing issues discussed in https://github.com/dagster-io/dagster/pull/19294

The main addition is `UPathIOManager.load_partitions` method, which should be considered a public interface for child IOManagers to override.  This interface is suited for `IOManager`s dealing with serialization formats which support partitioning out of the box. 

The benefits of this new interface are shown in `dagster-polars`, where a lot of boilerplate and work-around code has been removed, and new features are now supported thanks to `load_partitions`. 

:sparkles: Additions
 - Added `UPathIOManager.load_partitions` method. It's supposed to be overridden by child classes which implement special handling for loading multiple partitions. The default implementation loads them as a dict in a `for` loop (just like before). Child classes may want to change this. For example, for `DataFrame`s stored as `Parquet` datasets or `DeltaLake`, it makes sense to use `partition_filters` instead, thus loading multiple partitions in parallel. 
 - `dagster-polars`: now supports other data types rather than only string for the partitioning columns
 - `dagster-polars`: `PolarsDeltaIOManager` now supports `MultiPartitionsDefinition` with `DeltaLake` native partitioning. Metadata value `"partition_by": {"dim_1": "col_1", "dim_2": "col_2"}` should be specified to enable this feature. 

:heavy_minus_sign: Removed
 - removed the type check which forced the result of multiple partitions load to be a dictionary. Again, a single `DataFrame` can be loaded from multiple partitions, and the return object won't be a dict. 
 - removed a test which checked relevant behavior

:bug: Bug fixes
 - fixed a bug with `MultipartitionKey`'s backcompat_path in `UPathIOManager` - it was not used when loading multiple partitions
- `dagster-polars`: Previously, partition filtering happened after reading the whole `DeltaLake` table. This has been fixed.

:warning: Breaking changes
- `PolarsDeltaIOManager` no longer supports loading natively partitioned DeltaLake tables as dictionaries. They should be loaded as a single `pl.DataFrame`/`pl.LazyFrame` instead. 

## How I Tested These Changes
 - [x] existing tests
 - [x] using the new interface in `dagster-polars`
 - [x] added `MultiPartitionsDefinition` tests in `dagster-polars` 

Please let me know if you think this interface makes sense.

### CODE REVIEW

**Comment by danielgafni** [comment]
Re: `backcompat_path`

1. There was a typo in backcompat_path loading code: https://github.com/dagster-io/dagster/pull/21135/commits/5f82595938f108d1590e13ef94d587291751769d#diff-769da442e9a2e89f16a2942da633f014bab8aaeb01bb65ae53cba4f6c553baddR331
2. It's no longer a part of `_load_single_input`, because it's never used with partitions anymore

Actually, the `backcompat_path` was never used because of the typo! 

Should we just remove it to reduce complexity? Seems like there is no need to maintain it.

**Review by alangenfeld — COMMENTED** [review]
looks pretty reasonable on a first pass - can you address the pyright problems? https://buildkite.com/dagster/unit-tests/builds/613#018ecd8c-45b6-4d02-b207-8923858d1aa0

**Comment by danielgafni** [comment]
Hey @alangenfeld, done. 

P.S. something just broke in my local pyright installation, going to take some time to fix it. hopefully there are no new errors.

**Comment by danielgafni** [comment]
We can also consider adding base_path argument to load_partitions. It would receive the path to the directory containing partitions. 

Not sure if we need to do it or not. This path can be easily retrieved anyway.

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/upath_io_manager.py:296]: do you intend to resolve these TODOs before landing?

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/upath_io_manager.py:421]: can just drop this `TODO:` prefix if not planning to resolve

**Review by alangenfeld — APPROVED** [review]

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/storage/upath_io_manager.py:296]: Oh yeah, I'll find the exact version

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/storage/upath_io_manager.py:421]: Can you elaborate on what you mean? 

I'm not planning to do anything about this in this PR, I just left this comment for later

Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/storage/upath_io_manager.py:421]: I just meant you could drop the TODO: part of the comment 

I bias towards pushing to either tagging with an issue link if its something that should get addressed or leaving as a regular comment if it a vague idea of a future improvement with no clear intent of action. 

Just personal taste though, won't block merge on it.

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/storage/upath_io_manager.py:296]: i just merged it as is - can merge a PR that adds this if you find the exact version

**Pull request merged**
